File: gamedata/linebreak.py
#!/usr/bin/env python

# Copyright (c) 2014 SpinPunch. All rights reserved.
# Use of this source code is governed by an MIT-style license that can be
# found in the LICENSE file.

# Add linebreaks to a very big JSON file to make it more browser-friendly,
# without changing the interpretation of the JSON data in any way.

# C-accelerated version of this is available under linebreak/

import sys, cStringIO

class State(object):
    NORMAL=0
    QUOTE=1
    QUOTE_SPECIALCHAR=2
    SYMBOL=3
    NUMBER=4

if __name__ == '__main__':
    limit = 500
    src = sys.stdin
    src = cStringIO.StringIO(src.read()); src.seek(0)
    dst = sys.stdout
    buf = cStringIO.StringIO()
    state = State.NORMAL

    while True:
        c = src.read(1)
        if not c: break

        if state == State.NORMAL and buf.tell() >= limit:
            buf.write('\n')
            dst.write(buf.getvalue())
            buf = cStringIO.StringIO()

        if state == State.NORMAL and c in (' ', '\t', '\n', '\r'):
            continue

        buf.write(c)

        if state == State.NORMAL:
            if c == '"':
                state = State.QUOTE
            elif c.isalpha():
                state = State.SYMBOL
            elif (c.isdigit() or c == '.'):
                state = State.NUMBER
        elif state == State.SYMBOL:
            if not c.isalpha():
                state = State.NORMAL
        elif state == State.NUMBER:
            if not (c.isdigit() or c == '.' or c == 'e'):
                state = State.NORMAL
        elif state == State.QUOTE:
            if c == '\\':
                state = State.QUOTE_SPECIALCHAR
            elif c == '"':
                state = State.NORMAL
        elif state == State.QUOTE_SPECIALCHAR:
            state = State.QUOTE

    if state != State.NORMAL:
        raise Exception('ended in non-normal state')

    if buf.tell() > 0:
        tail = buf.getvalue()
        if tail != '\n':
            dst.write(tail)

    dst.write('\n')

File: gamedata/linebreak/linebreak.c
/* Break long lines in a JavaScript file, without changing its meaning at all. */
/* This should work identically to linebreak.py, just a lot faster. */

/* Copyright (c) 2015 SpinPunch. All rights reserved.
   Use of this source code is governed by an MIT-style license that can be
   found in the LICENSE file. */

#include <stdio.h>
#include <ctype.h>

enum State {
	NORMAL, QUOTE, QUOTE_SPECIALCHAR, SYMBOL, NUMBER
};

int main(int argc, char *argv[]) {
	static const int limit = 500; /* max output length (strings may break this) */
	enum State state = NORMAL;
	int len = 0; /* length of current line of output */
	int ends_with_linebreak = 0; /* true iff last character written was a linebreak */

	while(1) {
		char c = fgetc(stdin);
		if(c == EOF) { break; }

		if(state == NORMAL && len >= limit) {
			fputc('\n', stdout);
			len = 0;
			ends_with_linebreak = 1;
		}

		/* skip any non-significant whitespace from the input */
		if(state == NORMAL && isspace(c)) { continue; }

		/* write to output */
		fputc(c, stdout);
		len += 1;
		ends_with_linebreak = (c == '\n' ? 1 : 0);

		switch(state) {
		case NORMAL:
			if(c == '"') {
				state = QUOTE;
			} else if(isalpha(c)) {
				state = SYMBOL;
			} else if(isdigit(c) || c == '.') {
				state = NUMBER;
			}
			break;
		case SYMBOL:
			if(!isalpha(c)) {
				state = NORMAL;
			}
			break;
		case NUMBER:
			if(!isdigit(c) && c != '.' && c != 'e') {
				state = NORMAL;
			}
			break;
		case QUOTE:
			if(c == '\\') {
				state = QUOTE_SPECIALCHAR;
			} else if(c == '"') {
				state = NORMAL;
			}
			break;
		case QUOTE_SPECIALCHAR:
			state = QUOTE;
			break;
		}
	}
	if(state != NORMAL) {
		fprintf(stderr, "ended in non-normal state\n");
		return 1;
	}

	/* ensure the file ends with a (single) linebreak, even if source did not */
	if(!ends_with_linebreak) {
		fputc('\n', stdout);
	}

	return 0;
}
